### ISSUE TITLE
txid and p2p messages

### ISSUE THREAD

**Issue by zander** [bitjson/pmv3#1]
The readme has this snippet:

> When computing the transaction's hash for Outpoint Transaction Hash and for other P2P messages, any hashed witnesses are truncated from the preimage. (Hashed witness integrity is guaranteed by the witness hash in each respective input serialization.)

When I look specifically at P2P messages I want to point out a problem due to a design of the p2p layer being so integral to the data-structures.

A peer that receives a primitive on the p2p layer validates it and may reject it based on any consensus violation. The rejection of the data received from a peer is remembered in a 'recently-failed' list and subsequent INVs of that data are ignored.

This means that there is a strong relation between p2p-messages and primitive data structures, like transactions. The txid is used at both layers of the OSI stack to identify both the message and the primitive.

The result is that if you exclude a part of the message (like the witness) from the txid, there is no way to differentiate anymore if there was a corruption at the p2p layer or due to the basic primitive being invalid. This means that if a peer alters a correct transaction on the p2p layer, a honest peer has to decide to download other messages with the same txid, or reject all of them based on it not validating.

Tl;dr: we announce primitives using INV and the hash of the primitive and we ensure that the message-hash is identical to the primitive-hash in order to filter INVs based on both: 1) what we already know. 2) messages we rejected as invalid in the past.  
To not make those identical would mean it does not fit in the INV propagation method as is anymore.

**Comment by bitjson** [comment]
Hey @zander, thanks for opening an issue! This is a great point, and I'm afraid we don't have a great solution yet.

To confirm: you're considering a DOS where a malicious node modifies a valid transaction, changing the witness data appended to the end (and therefore making the hash invalid). When an honest node downloads this transaction, they find that it is invalid because the witness data does not match the expected hash. When they throw away the "invalid" transaction, they note the TXID as having been rejected in the past, and they stop attempting to download it from other peers.

Is that a good summary of the issue?

(For the benefit of anyone else reading this thread, a note: all P2P messages already include a 4-byte double-SHA256 checksum, so unintentional network corruption is easily differentiated from this attack.)

On one hand, I think it would be extremely valuable to avoid have two "types" of TXID (like BTC's `txid` and `wtxid`) – if we can figure out a strategy which only uses one ID (at least in "userland"), I think it will pay dividends in terms of developer and user experiences. However, maybe just for the P2P layer, implementations should use a sort of `P2P txid` which also includes the witness data. Ideally, we should keep the usage of this `ptxid` to only potentially-malicious P2P messages, since it would be a shame for it to complicate downstream experiences.

Another note: hashed witnesses should be relatively uncommon – very few types of transactions require them, and they cost 32-bytes per input to use, so other transactions are disincentivized from needlessly enabling them. (So at the very least, a relatively small fraction of all transactions would have different `txid` and `ptxid` values.)

Do you have any other ideas for making this work on the p2p layer?

**Comment by zander** [comment]
> Do you have any other ideas for making this work on the p2p layer?

To answer that we have to know the reason for this design in the first place. As I wrote on bitcoincashresearch, I'm unclear on the usage of the extra hash, without such knowledge I can't think along towards a solution.

More specific to this issue, you also have not given a reason for not including these hashes in the txid. Why do you omit these hashes from the txid calc?

**Comment by bitjson** [comment]
> Why do you omit these hashes from the txid calc?

@zander and I talked a bit more offline, but I also wanted to answer here for anyone else reading.

Using a hash of the unlocking bytecode for the transaction ID allows future transactions to inspect the contents of the current transaction without having to re-include its proofs. For contract-based tokens, this prevents transactions from growing in size over time. I just finished writing a full overview here: [Hashed Witnesses: Build Decentralized Applications on Bitcoin Cash](https://blog.bitjson.com/hashed-witnesses-build-decentralized-applications-on-bitcoin-cash-ccc2099a3956)

**Comment by bitjson** [comment]
After thinking about this more – I would really prefer that we avoid having multiple types of TXIDs, even on the P2P layer (this saves us from even needing to add another index mapping the two ID types to each other).

I think it's perfectly reasonable for network polices to handle malicious nodes dropping witness data: if a node receives a transaction which does not include witness data, it can safely assume the sending peer is "misbehaving". After enough infractions, ban the peer. 

There's no need to also ban the TXID (in case it's an otherwise valid transaction, and the malicious peer was just stripping the witness data from the message). I don't think banning the TXID can even save compute time if e.g. multiple malicious peers try to send the same "missing witness" transaction, so I think there is no advantage to including "missing witness" transactions in any sort of TX ban list.

So I think this is easy to solve with the current "misbehaving" peer controls.

**Comment by zander** [comment]
>  if a node receives a transaction which does not include witness data, it can safely assume the sending peer is "misbehaving". After enough infractions, ban the peer.

Thats a given.

> There's no need to also ban the TXID

The "recently rejected" map avoids re-downloading known invalid data. For instance a transaction that was meant for a different chain, and indeed also transactions that fail to meet our local policies and rules.

> I don't think banning the TXID can even save compute time

The resource it protects is the bandwidth. And in lesser amount the download queue (stuff stands in line for each peer).

> so I think there is no advantage to including "missing witness" transactions in any sort of TX ban list.

The advantage is protecting the peer from being DOSed.

I'm willing to see how this plays out for now, although its a deviation from the norm and requires an exception in generic code which just feels bad.

Also you might need to convince BU to re-instate the per-message-checksum.

**Comment by bitjson** [comment]
> [...] although its a deviation from the norm and requires an exception in generic code which just feels bad.

Ah, I see. I suppose that's a decision each implementation will have to make. Hopefully people will come up with clean solutions.

> Also you might need to convince BU to re-instate the per-message-checksum.

Thanks for the heads up! I didn't realize they [removed P2P message checksum validation](https://github.com/BitcoinUnlimited/BitcoinUnlimited/commit/35c662e64).

**Comment by rnbrady** [comment]
Although the `txid` is calculated differently in PMv3, it's still a 32 byte digest _covering the entire transaction_ so it seems ok for use as an inventory identifier. 

In BCHN the P2P layer is already [deserialising the transaction and using the higher level TXID for inventory](https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/blob/master/src/net_processing.cpp#L2844-2849):

```
        CTransactionRef ptx;
        vRecv >> ptx;                     // deserialise
        const CTransaction &tx = *ptx;
        const TxId &txid = tx.GetId();    // get txid 

        CInv inv(MSG_TX, txid);           // use txid as inventory id
```

Or is there another layer below that?

> The result is that if you exclude a part of the message (like the witness) from the txid, there is no way to differentiate anymore if there was a corruption at the p2p layer or due to the basic primitive being invalid.

PMv3 doesn't exclude the witnesses from the `txid` completely. It excludes the raw witnesses but includes a hash of each one. The `txid` will become more complicated to calculate but integrity is not lost.

It is necessary to deserialise for two reasons:
1. To work out where to truncate the raw transaction that is hashed it to obtain to the `txid`.
2. To check the hashes before the truncation point match the witnesses after it. 

The second step is not required to calculate the `txid` but must be performed to ensure integrity.

**Comment by zander** [comment]
> Although the txid is calculated differently in PMv3, it's still a 32 byte digest covering the entire transaction so it seems ok for use as an inventory identifier.

Now Jason is back, I'd like to know if this is true.

I was under the impression that if a witness is altered that the txid does NOT change. Which would make the above statement false.

Would like to know.

**Comment by bitjson** [comment]
Hey @rnbrady and @zander – sorry I missed responding here before!

Yes, @rnbrady is correct, the witness (now "detached proof") is covered by the TXID, it's simply hashed before being included in the TXID preimage. Sorry I must have misunderstood your question initially @zander. Thanks for recognizing @rnbrady!

In the latest revision, I tried to make this point more clear, and I also added a note about blacklisting TXIDs (just in case anyone is doing that):

> Note: node implementations must never blacklist TXIDs for non-presence of detached proof(s), but should ban peers which repeatedly transmit malformed transactions (such as transactions from which detached proofs have been dropped).
